### ISSUE TITLE
Improve Model Relationship Guides

### ISSUE THREAD

**Issue by rasmus1610** [phoenixframework/phoenix_guides#607]
I had a huge problem finding a proper example of how to build associations for models in the controller (build_assoc/3 and stuff). Defining the relationships in the model wasn't the problem but how to build the changeset with the associations was actually pretty hard for me as a newbie.

An improved guide on this topic would be awesome. It doesn't really cover it right now.

I'm happy to contribute to it, but I'd like some help, cause I'm not really sure I understand Phoenix good enough to write guides for it.

**Comment by willykaram** [comment]
**_TL;DR: It appears a Phoenix guide on this could be beneficial. I can help @rasmus1610 create it you want to, or vice versa I could create it and you could review my draft. If you're interested, could you perhaps some more details here on the problem you ran into, and the resolution). I've posted some similar issues people have run into below._**

@rasmus1610 I've done some recent research on this issue, as someone I was helping encountered a similar issue the last week or so (and I actually got stumped while helping them too).

I think that the Associations guide that @josevalim recently updated does a pretty good job of covering the Ecto side of things:

https://github.com/elixir-ecto/ecto/blob/master/guides/Associations.md

Additionally, the many_to_many chapter in the What's New in Ecto 2.0 book and a related blog posts are also good resources:

http://blog.plataformatec.com.br/wp-content/uploads/2017/05/whats-new-in-ecto-2-1.pdf

http://blog.plataformatec.com.br/2016/12/many-to-many-and-upserts/

http://blog.plataformatec.com.br/2015/08/working-with-ecto-associations-and-embeds/

Those materials seem to do a good job of getting everyone through creating their model, schema, migrations, and interacting with the model in IEX. 

However, from what I've seen on multiple threads, there seems to be quite a bit of confusion in the community as people try to implement that in Pheonix -- particularly when [implementing in forms](https://hexdocs.pm/phoenix_html/Phoenix.HTML.Form.html#multiple_select/4). This is likely in part due to many_to_many associations being added to Ecto 2.0 (as well as some of the more recent changes to model best practices with Phoenix 1.3). For instance:

https://elixirforum.com/t/adding-many-to-many-assoc-on-insert/1233

https://elixirforum.com/t/help-with-building-associations/3653/12

https://elixirforum.com/t/unexpected-behaviour-with-many-to-many-nested-association-casting/4284/

So with that in mind I suggest that we write up a guide, along with some sample code or a reference app, which covers this. I would suggest that we simply pick-up where either Elixir Associations guide leaves off and/or the What's New in Ecto 2.0 Book leaves off -- i.e., with a Phoenix implementation of the Blog/Tags or ToDo examples. From what I've seen online there doesn't appear to be an example online ATM that's complete, simple and current (Ecto 2.0 & Phx 1.3).

**Comment by ryanwinchester** [comment]
Associations has been one of my biggest pain points while transitioning to phoenix.

I was constantly in the Ecto docs reading and re-reading the `cast_assoc` and `put_assoc` sections and trying and retrying different ways of doing associations and changesets.

**Comment by willykaram** [comment]
@ryanwinchester thanks for the feedback. That consistent with what I've heard from a few folks recently, so I'm thinking that the guide that @rasmus1610 has proposed would essentially cover by extending on the existing cast_/put_assoc content that @josevalim while also providing code/examples for each type of association (e.g., a simple blog site with users, comments, tags, and  related posts should enable us to cover each use case).

If you have any specific challenges and/or solutions it'd be great to post them/Gists here, and we can ensure they're covered in the guide/examples. If anyone has any other links in addition to the above that we can build upon, that'd be great too, such as:

- [Adding Comments with Ecto Associations](https://elixircasts.io/adding-comments-with-ecto-associations)
- [Nested Routes and Associations - Elixir and Phoenixphoenix/](http://danbickford007.com/blog/2016/05/13/nested-routes-and-associations-elixir-and-)
- [Preloading Nested Associations with Ecto](https://robots.thoughtbot.com/preloading-nested-associations-with-ecto)

While this can often be tricky with other technologies, we're very fortunate that the current BEAM stack provides some great solutions; so I think it's just a question of us/community members chipping in on the documentation front the help the core team -- which I'm happy to do on this issue, as it's a topic we likely need to cover as part of helping @chrismccord with the Phoenix 1.3 Guides Overhaul, #680.

**Comment by josevalim** [comment]
Great reply @willykaram. I agree that getting some of the knowledge in the book and in our blog posts can be helpful and show a "full stack" functioning of associations compared to the Ecto guide. The Phoenix team is free to use any of those materials. Linking to the Ecto book is probably beneficial too. :)

**Comment by Gazler** [comment]
I think links out to the existing Ecto documentation is a good idea, since it means we don't have to do maintainance on both sets of guides. We should be sure to link to a specific version of the guides too. I think what we need to cover is Ecto and associations in the context of...uhhh...contexts.

@josevalim Are there any plans to publish the Ecto guides on hexdocs?

**Comment by josevalim** [comment]
Guides will be out on the next release. Maybe in 2 months?
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by willykaram** [comment]
Thanks @gazler @josevalim for the feedback. 

I agree with @gazler that the aim should be to keep the guides as minimal as possible. I also wasn't sure whether the proceed since in #680 @chrismccord mentioned that WRT the Ecto Model guides the plan was to "_not discuss relationships in detail, defering to context guide which will build off this example_". 

That said, I do think it would make sense to have a bonus guide that goes into detail on full stack implementations of associations in Pheonix (particularly for 1.3, w/the removal of the models folder). Or alternatively to have a blog post somewhere that cover the full stack Phoenix use case.

Although some of the existing resources are quite good (e.g., [Ecto Associations guide](https://github.com/elixir-ecto/ecto/blob/master/guides/Associations.md) and [Many to many and upserts blog post/book chapter](http://blog.plataformatec.com.br/2016/12/many-to-many-and-upserts/)), they don't translate perfectly to Phoenix.

For instance, I was starting out by trying to replicate the app/solution outlined in the [blog post](http://blog.plataformatec.com.br/2016/12/many-to-many-and-upserts/), which recommends piping the tags list through [String.split/3](https://hexdocs.pm/elixir/String.html#split/3). However, OOB this approach doesn't necessarily work in Pheonix if the tags are being passed in from a [multiple select](https://hexdocs.pm/phoenix_html/Phoenix.HTML.Form.html#multiple_select/4) helper on the edit form, since there's an opening/closing square bracket and quotes  (i.e., something like ["2","3"]) (unless there's another preferred approach for handling passing the multiple_select from the form). You can download/run my initial implementation from [this repo](https://github.com/willykaram/phx-many-to-many-and-upserts) (sorry didn't have a chance to deploy it online yet).

There also seem to be a number of posts [like this](https://elixirforum.com/t/association-help/5851) on the forums or Stack Overflow, similar to this issue, where people seem to grasp how to setup associations in their schema/migrations (and testing in IEX), but they are struggling to implement them when gathering input from the user with Phoenix and then passing it back to their model/controller and Ecto for CRUD operations. This is pretty essential functionality that people seem to be getting stuck on, so that's why I was suggesting a guide or more documentation might be helpful (ideally along with a clear reference Phoenix implementation of the various approaches to associations so people can setup/see a working example).

**Comment by willykaram** [comment]
In the interest of clarity/precision, the intent wouldn't be to duplicate the Ecto associations guide or other content, no to cover context or application design. Rather, given that that there are multiple ways to convert this sort of user input/strings to a list of maps for purposes of Ecto many_to_many associations and persistence, the intent would simply be to connect those dots for people. Specifically covering the relevant Phoenix considerations relating to implementing this in Phoenix controllers and templates/helpers. Of all the post/examples online, I wasn't able to find a simple complete/simple one to point someone too (unless I missed it). If you think that's needed/makes sense I can create a draft of the guide. If not, I can create blog post instead.

I realize we have lots of other work to do for the 1.3 guides overhaul, so don't want to unnecessarily add to that workload (which I'm also trying to help with ATM).

**Comment by chrismccord** [comment]
The context guide will introduce relationships and show examples so we may be able to handle it there. I suggest we wait until I complete the context guide and see if there are still gaps to fill Ecto relationship wise.

**Comment by jeregrine** [comment]
@rasmus1610 @chrismccord do you believe the new guide covers the bases enough to close this one?

**Comment by willykaram** [comment]
@jeregrine I'm working on a community contributed guide on this that would reside outside the official guides, and which will cover these details. Ideally I'll have it ready tomorrow at the sprint so I think we could close the issue.
